File: src/ruby/nodes/UnaryExpr.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'treetop'
require_relative 'BooleanExpr'

# Variable interpolation in a template
class UnaryExpr < BooleanExpr
  def truthy?(context)
    if negated.text_value.length.positive?
      !elements[1].val.truthy?(context)
    else
      elements[1].val.truthy?(context)
    end
  end
end

File: cpp/src/nodes/unary_expr.cpp
#include "unary_expr.hpp"

UnaryExpr::UnaryExpr(bool negated, BooleanPtr expr) : negated(negated), expr(expr) {}

bool UnaryExpr::truthy(Json &context) const {
  if (negated) {
    return !expr->truthy(context);
  } else {
    return expr->truthy(context);
  }
}
